aquadoggo/src/graphql/schema.rs

```diff
@@ -39,3 +47,143 @@ pub fn build_root_schema(store: SqlStorage, tx: ServiceSender) -> RootSchema {
+/// content.
+async fn build_schema_with_workaround(shared: GraphQLSharedData) -> RootSchema {
+    // Store temporary JSON file with all serialised application schemas from database.
+    //
+    // @TODO: We could try to still access the database by creating a static interface to it?
+    let all_schemas = shared
+        .schema_service
+        .all_schemas()
+        .await
+        .expect("Loading all schemas from database failed");
+    let temp_file = TempFile::save(&all_schemas, TEMP_FILE_PATH);
+
+    // Build the actual GraphQL root schema, this will internally read the created JSON file and
+    // accordingly build the schema
+    let schema = build_root_schema(shared.store, shared.tx, shared.schema_service);
+
+    // Remove temporary file as we don't need it anymore
+    temp_file.unlink();
+
+    schema
+}
+
+/// List of created GraphQL root schemas.
+type GraphQLSchemas = Arc<Mutex<Vec<RootSchema>>>;
+
+/// Shared types between GraphQL schemas.
+#[derive(Clone, Debug)]
+pub struct GraphQLSharedData {
+    /// Database interface.
+    store: SqlStorage,
+
+    /// Communication bus interface to send messages to other services.
+    tx: ServiceSender,
+
+    /// Schema provider giving us access to currently known schemas.
+    schema_service: SchemaService,
+}
+
+/// Builds new GraphQL schemas dynamically and executes the latest GraphQL schema for incoming
+/// queries.
+///
+/// This manager allows us to introduce new GraphQL schemas during runtime as it internally handles
+/// a list of schemas (behind a mutex) and automatically picks the "latest" as soon as a query
+/// needs to be executed.
+///
+/// With this we can easily add "new" schemas to the list in the background while current queries
+/// still get processed using the "old" schema.
+#[derive(Clone)]
+pub struct GraphQLSchemaManager {
+    /// List of all built GraphQL root schemas.
+    schemas: GraphQLSchemas,
+
+    /// Commonly shared types for GraphQL schemas.
+    shared: GraphQLSharedData,
+}
+
+impl GraphQLSchemaManager {
+    /// Returns a new instance of `GraphQLSchemaManager`.
+    pub async fn new(store: SqlStorage, tx: ServiceSender, schema_service: SchemaService) -> Self {
+        let schemas = Arc::new(Mutex::new(Vec::new()));
+        let shared = GraphQLSharedData {
+            store,
+            tx,
+            schema_service,
+        };
+
+        // Create manager instance and spawn internal watch task
+        let manager = Self { schemas, shared };
+        manager.spawn_schema_added_task().await;
+
+        manager
+    }
+
+    /// Subscribes to `SchemaService` for newly added schemas.
+    ///
+    /// This spawns a task which listens to new p2panda schemas to accordingly build a GraphQL
+    /// schema which will be added to the list.
+    // @TODO: This manager does not "clean up" outdated schemas yet, they will just be appended to
+    // an ever-growing list.
```

Comment:
:point_up: Thats of course a potential source of ever-growing memory the longer the node runs, but I don't see a need to fix this in this PR.

I believe the mutex protects us from removing schemas from the list _while_ they are being executed somewhere else due to a pending request. So it should be safe to just remove items?

Reply:
What if the schema was used in a way that involves multiple locks during one process? Like a subscription request, could that involve a series of locks?

Reply:
Yeah, subscriptions will need also some special interface between but I don't think its an issue. We will just have to do something similar as we do with the `GraphQLSchemaManager`, aka, something which abstracts away that there is actually a changing schema behind.